api/client/trust.go

```diff
@@ -0,0 +1,437 @@
+package client
+
+import (
+	"bufio"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/registry/client/auth"
+	"github.com/docker/distribution/registry/client/transport"
+	"github.com/docker/docker/cliconfig"
+	"github.com/docker/docker/pkg/ansiescape"
+	"github.com/docker/docker/pkg/ioutils"
+	flag "github.com/docker/docker/pkg/mflag"
+	"github.com/docker/docker/pkg/tlsconfig"
+	"github.com/docker/docker/registry"
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/pkg/passphrase"
+	"github.com/endophage/gotuf/data"
+)
+
+var untrusted bool
+
+func addTrustedFlags(fs *flag.FlagSet, verify bool) {
+	var trusted bool
+	if e := os.Getenv("DOCKER_TRUST"); e != "" {
+		var err error
+		trusted, err = strconv.ParseBool(e)
+		if err != nil {
+			// treat any other value as true
+			trusted = true
+		}
+	}
+	message := "Skip image signing"
+	if verify {
+		message = "Skip image verification"
+	}
+	fs.BoolVar(&untrusted, []string{"-untrusted"}, !trusted, message)
+}
+
+func isTrusted() bool {
+	// TODO(dmcgowan): false if name is hex image id
+	// TODO(dmcgowan): whitelist specific names to be untrusted
+	// TODO(dmcgowan): canonicalize name before checking whitelist
+	return !untrusted
+}
+
+var targetRegexp = regexp.MustCompile(`([\S]+): digest: ([\S]+) size: ([\d]+)`)
+
+type target struct {
+	reference registry.Reference
+	digest    digest.Digest
+	size      int64
+}
+
+func (cli *DockerCli) trustDirectory() string {
+	return filepath.Join(cliconfig.ConfigDir(), "trust")
+}
+
+// certificateDirectory returns the directory containing
+// TLS certificates for the given server. An error is
+// returned if there was an error parsing the server string.
+func (cli *DockerCli) certificateDirectory(server string) (string, error) {
+	u, err := url.Parse(server)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(cliconfig.ConfigDir(), "tls", u.Host), nil
+}
+
+func trustServer(index *registry.IndexInfo) string {
+	if s := os.Getenv("DOCKER_TRUST_SERVER"); s != "" {
+		if !strings.HasPrefix(s, "http") {
+			return "https://" + s
+		}
```

Comment:
This was actually our intention from day one.

Since we are using TOFU, we have to enforce docker-cli <-> notary-server communication using HTTPS, or a MiTM could very easily server a inject a fake certificate the first time a user connects.

This way we are using the public PKI to protect the TOFU (trust on first use). We're calling it: TOFUS!